File: GUI Objects/FoodListTableDropSupport.m
#import "FoodListTableDropSupport.h"
#import "Food.h"
#import "FoodList.h"
#import "FoodDragSupportBase.h"
#import "GUIController.h"

@implementation FoodListTableDropSupport

//Need to implement data source routines so we include them
- (int)numberOfRowsInTableView:(NSTableView *)aTableView 
   {
    return 0; // return 0 so the table view will fall back to getting data from its binding
   }



- (id)tableView:(NSTableView *)aTableView objectValueForTableColumn:(NSTableColumn *)aTableColumn row:(int)rowIndex 
   {
    return nil;  // return nil so the table view will fall back to getting data from its binding 
   }



- (NSDragOperation)tableView:(NSTableView *)tableView validateDrop:(id <NSDraggingInfo>)info proposedRow:(int)row proposedDropOperation:(NSTableViewDropOperation)operation
   {
   
   //Only allow dropping on lists not between
   if ( operation == NSTableViewDropOn )
      {
      return NSDragOperationCopy;
      }
   else
      {
      return NSDragOperationNone;
      }
   }



- (BOOL)tableView:(NSTableView *)tableView acceptDrop:(id <NSDraggingInfo>)info row:(int)row dropOperation:(NSTableViewDropOperation)operation 
   {
	BOOL success = NO;
   
   NSDictionary *infoForBinding = [tableView infoForBinding:NSContentBinding];

   if (infoForBinding == nil) 
      { return( NO ); }
   
   //Get array of all food lists
   NSArrayController *arrayController = [infoForBinding valueForKey:NSObservedObjectKey];

   //Get our food list
   FoodList* FoodListToModify = [[arrayController arrangedObjects] objectAtIndex:row];

   if ( FoodListToModify == nil )
      { return( NO ); }
      
   //Food list internal set
   NSMutableSet* FoodsSet = [FoodListToModify mutableSetValueForKey:@"foods"];
      
   NSManagedObjectContext *context = [ApplicationDelegate managedObjectContext];

   if ( context == nil )
      { return( NO ); }
     

    // Get object URIs from paste board
    NSData* data = [[info draggingPasteboard] dataForType:FoodDataDragType];
    NSArray* objectURIs = [NSUnarchiver unarchiveObjectWithData:data];
    if (!objectURIs) 
      { return NO; }
    
   NSEnumerator* enumerator = [objectURIs objectEnumerator];

   NSURL* currentObjectURI;

    while (currentObjectURI = [enumerator nextObject]) 
      {
      
      // Get managed object
      NSManagedObjectID* objectID = [[context persistentStoreCoordinator]
                                          managedObjectIDForURIRepresentation:currentObjectURI];

      Food* FoodObjectExtracted = (Food*) [context objectWithID:objectID];
      if ( FoodObjectExtracted == NULL )
         { continue; }
         
      [ FoodsSet addObject: FoodObjectExtracted ];

      success = YES;
        
      }
       
      [self.myTableView reloadData];
   return( success );
       
   }


@end

File: GUI Objects/FoodListTableDropSupport.swift
//
//  FoodListTableDropSupport.swift
//  BeFit
//
//  Created by Andrej Slegl on 11/5/19.
//

import Foundation
import Cocoa

class FoodListTableDropSupport: NSObject, NSTableViewDelegate, NSTableViewDataSource {
    @IBOutlet var ApplicationDelegate: BeFit_AppDelegate!
    @IBOutlet var myTableView: NSTableView!
    
    func numberOfRows(in tableView: NSTableView) -> Int {
        return 0 // return 0 so the table view will fall back to getting data from its binding
    }
    
    func tableView(_ tableView: NSTableView, objectValueFor tableColumn: NSTableColumn?, row: Int) -> Any? {
        return nil // return nil so the table view will fall back to getting data from its binding
    }
    
    func tableView(_ tableView: NSTableView, validateDrop info: NSDraggingInfo, proposedRow row: Int, proposedDropOperation dropOperation: NSTableView.DropOperation) -> NSDragOperation {
        switch dropOperation {
        case .on:
            return .copy
        default:
            return []
        }
    }
    
    func tableView(_ tableView: NSTableView, acceptDrop info: NSDraggingInfo, row: Int, dropOperation: NSTableView.DropOperation) -> Bool {
        guard let infoForBinding = tableView.infoForBinding(.content) else {
            return false
        }
        
        guard let arrayController = infoForBinding[.observedObject] as? NSArrayController else {
            return false
        }
        
        let arrangedObjects = arrayController.arrangedObjects as! [FoodList]
        
        guard row < arrangedObjects.count else {
            return false
        }
        
        let foodListToModify = arrangedObjects[row]
        
        guard let context = ApplicationDelegate?.managedObjectContext else {
            return false
        }
        
        let data = info.draggingPasteboard.data(forType: .foodDataDragType) ?? Data()
        guard let objectURIs = NSUnarchiver.unarchiveObject(with: data) as? [URL] else {
            return false
        }
        
        for currentObjectURI in objectURIs {
            guard let objectID = context.persistentStoreCoordinator?.managedObjectID(forURIRepresentation: currentObjectURI) else {
                continue
            }
            
            guard let foodObject = context.object(with: objectID) as? Food else {
                continue
            }
            
            foodListToModify.addToFoods(foodObject)
        }
        
        myTableView.reloadData()
        return true
    }
}
